Objects/unionobject.c

```diff
@@ -328,7 +328,12 @@ union_getitem(PyObject *self, PyObject *item)
     unionobject *alias = (unionobject *)self;
     // Populate __parameters__ if needed.
     if (alias->parameters == NULL) {
-        alias->parameters = _Py_make_parameters(alias->args);
+        Py_BEGIN_CRITICAL_SECTION(alias);
+        if (alias->parameters == NULL) {
```

Comment:
There are now two levels of `alias->parameters == NULL` checks. Is this right?

Reply:
It's deliberate. I added a comment. Is the code less surprising with the comment?

```c
        // Need to check again once we got the lock, another thread may have
        // initialized it while we were waiting for the lock.
```

Reply:
I read again your review, sorry I misunderstood. In fact, it's a micro-optimization: it avoids the "slow" critical section if *parameters* is already initialized. Do you think that it's not worth it?

Reply:
No, it's not safe. To make it safe, we need to use an atomic load (outside the critical section) and an atomic store when writing to `alias->parameters`.

I think there's also another lazy initialization of `parameters` in a function below.

Reply:
> we need to use an atomic load (outside the critical section) and an atomic store when writing to alias->parameters

Oh ok, I updated my PR for that.

> I think there's also another lazy initialization of parameters in a function below.

Well spotted, I also updated this function (union_parameters()).

Reply:
I removed the optimization. I don't think that it's worth it.